kms_log/main.tf

```diff
@@ -558,4 +593,469 @@ resource "aws_cloudwatch_metric_alarm" "dead_letter" {
+            SNS_EVENT_TOPIC_ARN = "${aws_sns_topic.kms_logging_events.arn}"
+        }
+    }
+
+    tags {
+        source_repo = "https://github.com/18F/identity-lambda-functions"
+        environment = "${var.env_name}"
+    }
+}
+
+resource "aws_lambda_event_source_mapping" "cloudtrail_processor" {
+    count = "${var.kmslogging_service_enabled}"
+    event_source_arn = "${aws_sqs_queue.kms_ct_events.arn}"
+    function_name = "${aws_lambda_function.cloudtrail_processor.arn}"
+}
+
+
+data "aws_iam_policy_document" "ctprocessor_cloudwatch" {
+    statement {
+        sid = "CreateLogGroup"
+        effect = "Allow"
+        actions = [
+            "logs:CreateLogGroup"
+        ]
+
+        resources = [
+            "arn:aws:logs:${var.region}:${data.aws_caller_identity.current.account_id}:*",
+        ]
+    }
+    statement {
+        sid = "PutLogEvents"
+        effect = "Allow"
+        actions = [
+            "logs:CreateLogStream",
+            "logs:PutLogEvents"
+        ]
+
+        resources = [
+            "arn:aws:logs:${var.region}:${data.aws_caller_identity.current.account_id}:log-group:/aws/lambda/${aws_lambda_function.cloudtrail_processor.id}:*"
+        ]
+    }
+}
+
+data "aws_iam_policy_document" "ctprocessor_kms" {
+    statement {
+        sid = "KMS"
+        effect = "Allow"
+        actions = [
+            "kms:Decrypt",
+            "kms:GenerateDataKey"
+        ]
+
+        resources = [
+            "${aws_kms_key.kms_logging.arn}"
+        ]
+    }
+}
+
+data "aws_iam_policy_document" "ctprocessor_dynamodb" {
+    statement {
+        sid = "DynamoDb"
+        effect = "Allow"
+        actions = [
+            "dynamodb:PutItem",
+            "dynamodb:GetItem"
+        ]
+
+        resources = [
+            "${aws_dynamodb_table.kms_events.arn}"
+        ]
+    }
+}
+
+data "aws_iam_policy_document" "ctprocessor_sns" {
+    statement {
+        sid = "SNS"
+        effect = "Allow"
+        actions = [
+            "sns:Publish"
+        ]
+
+        resources = [
+            "${aws_sns_topic.kms_logging_events.arn}"
+        ]
+    }
+}
+
+data "aws_iam_policy_document" "ctprocessor_sqs" {
+    statement {
+        sid = "SQS"
+        effect = "Allow"
+        actions = [
+            "sqs:DeleteMessage",
+            "sqs:ChangeMessageVisibility",
+            "sqs:ReceiveMessage",
+            "sqs:SendMessage",
+            "sqs:GetQueueAttributes"
+        ]
+
+        resources = [
+            "${aws_sqs_queue.kms_ct_events.arn}"
+        ]
+    }
+}
+
+data "aws_iam_policy_document" "assume-role" {
+    statement {
+        actions = [
+            "sts:AssumeRole"
+        ]
+        principals {
+            type = "Service"
+            identifiers = [
+                "lambda.amazonaws.com"
+            ]
+        }
+    }
+}
+
+resource "aws_iam_role" "cloudtrail_processor" {
+    name = "${aws_lambda_function.cloudtrail_processor.id}-execution"
```

Comment:
In my version of Terraform this is a cycle, because the lambda refers to the role and the role to the Lambda. Do I need to upgrade?

Reply:
I created a local variable for the lambda name so it can be used for both the role and lambda.